File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/LoaderConfiguration.java
package ru.kmorozov.library.data.loader;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Lazy;
import ru.kmorozov.gbd.core.logic.connectors.HttpConnector;
import ru.kmorozov.gbd.core.logic.connectors.google.GoogleHttpConnector;
import ru.kmorozov.gbd.core.logic.connectors.webdriver.WebDriverHttpConnector;
import ru.kmorozov.gbd.logger.Logger;
import ru.kmorozov.onedrive.client.OneDriveProvider;
import ru.kmorozov.onedrive.client.authoriser.AuthorisationProvider;
import ru.kmorozov.onedrive.client.authoriser.TokenFactory;
import ru.kmorozov.onedrive.client.exceptions.InvalidCodeException;

import java.io.File;
import java.io.IOException;
import java.net.URL;

/**
 * Created by km on 26.12.2016.
 */

@Configuration
public class LoaderConfiguration {

    private static final Logger logger = Logger.getLogger(LoaderConfiguration.class);

    @Value("${onedrive.key}")
    public String oneDriveKeyFileName;

    @Value("${webdriver.chrome.driver}")
    public String webdriverChromeDriverPath;

    @Value("${library.http.connector.type}")
    public String httpConnectorType;

    @Value("${onedrive.user}")
    public String oneDriveUser;

    @Value("${onedrive.password}")
    public String oneDrivePassword;

    @Value("${onedrive.clientId}")
    public String onedriveClientId;

    @Value("${onedrive.clientSecret}")
    public String onedriveClientSecret;

    @Bean @Lazy
    public OneDriveProvider api() {
        final URL keyResource = this.getClass().getClassLoader().getResource(this.oneDriveKeyFileName);

        if (keyResource == null) {
            LoaderConfiguration.logger.warn("Key file not found, creating new...");
        }

        File keyFile = new File(keyResource == null ? this.oneDriveKeyFileName : keyResource.getFile());

        if (!keyFile.exists()) {
            try {
                keyFile.createNewFile();
            } catch (final IOException e) {
                LoaderConfiguration.logger.error("Create keyFile error", e);
            }
        }

        AuthorisationProvider authoriser = null;

        try {
            authoriser = AuthorisationProvider.FACTORY.create(keyFile.toPath(), this.onedriveClientId, this.onedriveClientSecret);
        } catch (InvalidCodeException cee) {
            System.setProperty("webdriver.chrome.driver", this.webdriverChromeDriverPath);
            if (TokenFactory.generateToken(keyFile, this.oneDriveUser, this.oneDrivePassword, this.onedriveClientId))
                try {
                    authoriser = AuthorisationProvider.FACTORY.create(keyFile.toPath(), this.onedriveClientId, this.onedriveClientSecret);
                } catch (IOException e) {
                    LoaderConfiguration.logger.error("OneDrive API init error", e);
                }
            else
                throw new RuntimeException(cee);
        } catch (IOException e) {
            LoaderConfiguration.logger.error("OneDrive API init error", e);
        }

        return OneDriveProvider.FACTORY.readWriteApi(authoriser);
    }

    @Bean
    @Lazy
    public HttpConnector getConnector() {
        switch (this.httpConnectorType) {
            case "chrome":
                System.setProperty("webdriver.chrome.driver", this.webdriverChromeDriverPath);
                return new WebDriverHttpConnector();
            case "google":
                return new GoogleHttpConnector();
            default:
                return new GoogleHttpConnector();
        }
    }
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/LoaderConfiguration.kt
package ru.kmorozov.library.data.loader

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Lazy
import ru.kmorozov.gbd.core.logic.connectors.HttpConnector
import ru.kmorozov.gbd.core.logic.connectors.google.GoogleHttpConnector
import ru.kmorozov.gbd.logger.Logger
import ru.kmorozov.onedrive.client.OneDriveProvider
import ru.kmorozov.onedrive.client.OneDriveProvider.FACTORY
import ru.kmorozov.onedrive.client.authoriser.AuthorisationProvider
import ru.kmorozov.onedrive.client.authoriser.TokenFactory
import ru.kmorozov.onedrive.client.exceptions.InvalidCodeException
import java.io.File
import java.io.IOException

/**
 * Created by km on 26.12.2016.
 */

@Configuration
@ComponentScan(basePackageClasses = arrayOf(OneDriveProvider::class))
open class LoaderConfiguration {

    @Value("\${onedrive.key}")
    var oneDriveKeyFileName: String? = null

    @Value("\${webdriver.chrome.driver}")
    var webdriverChromeDriverPath: String? = null

    @Value("\${library.http.connector.type}")
    var httpConnectorType: String? = null

    @Value("\${onedrive.user}")
    var oneDriveUser: String? = null

    @Value("\${onedrive.password}")
    var oneDrivePassword: String? = null

    @Value("\${onedrive.clientId}")
    var onedriveClientId: String? = null

    @Value("\${onedrive.clientSecret}")
    var onedriveClientSecret: String? = null

    open val connector: HttpConnector
        @Bean
        @Lazy
        get() {
            when (httpConnectorType) {
                "google" -> return GoogleHttpConnector()
                else -> return GoogleHttpConnector()
            }
        }

    @Bean
    @Lazy
    open fun api(): OneDriveProvider {
        val keyResource = javaClass.classLoader.getResource(oneDriveKeyFileName!!)

        if (keyResource == null) {
            logger.warn("Key file not found, creating new...")
        }

        val keyFile = File(if (keyResource == null) oneDriveKeyFileName else keyResource.file)

        if (!keyFile.exists()) {
            try {
                keyFile.createNewFile()
            } catch (e: IOException) {
                logger.error("Create keyFile error", e)
            }

        }

        var authoriser: AuthorisationProvider? = null

        try {
            authoriser = AuthorisationProvider.FACTORY.create(keyFile.toPath(), onedriveClientId!!, onedriveClientSecret!!)
        } catch (cee: InvalidCodeException) {
            System.setProperty("webdriver.chrome.driver", webdriverChromeDriverPath!!)
            if (TokenFactory.generateToken(keyFile, oneDriveUser!!, oneDrivePassword!!, onedriveClientId!!))
                try {
                    authoriser = AuthorisationProvider.FACTORY.create(keyFile.toPath(), onedriveClientId!!, onedriveClientSecret!!)
                } catch (e: IOException) {
                    logger.error("OneDrive API init error", e)
                }
            else
                throw RuntimeException(cee)
        } catch (e: IOException) {
            logger.error("OneDrive API init error", e)
        }

        return FACTORY.readWriteApi(authoriser!!)
    }

    companion object {

        private val logger = Logger.getLogger(LoaderConfiguration::class.java)
    }
}
